teps/0118-matrix-with-explicit-combinations-of-parameters.md

```diff
@@ -0,0 +1,415 @@
+
+    - IMAGE: "image-3"
+      DOCKERFILE: "path/to/Dockerfile3"
+```
+
+The `Pipeline` addressing this use case will be defined as shown below:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: pipeline-to-build-images-from-a-single-repo
+spec:
+  workspaces:
+    - name: shared-workspace
+  tasks:
+    - ...
+    - name: kaniko-build
+      taskRef:
+        name: kaniko
+      workspaces:
+        - name: source
+          workspace: shared-workspace
+      matrix:
+        include:
+          - name: build-1
+            params:
+              - name: IMAGE
+                value: "image-1"
+              - name: DOCKERFILE
+                value: "path/to/Dockerfile1"
+          - name: build-2
+            params:
+              - name: IMAGE
+                value: "image-2"
+              - name: DOCKERFILE
+                value: "path/to/Dockerfile2"
+          - name: build-3
+            params:
+              - name: IMAGE
+                value: "image-3"
+              - name: DOCKERFILE
+                value: "path/to/Dockerfile3"                
+    - ...
+```
+
+This configuration allows users to take advantage of `Matrix` to fan out without having an auto-populated `Matrix`.
+
+## Alternatives
+
+### Using Object Parameters
+
+The explicit combinations of `Parameters` could be of type `Object`.
+
+Instead of the [the solution](#include-specific-combinations-in-the-matrix) for including a specific combination in the
+`Matrix`, we could use `Parameter` of type `Object` as such:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: pipeline-to-build-and-test-go-project
+spec:
+  workspaces:
+    - name: shared-workspace
+  tasks:
+    - ...
+    - name: golang-test
+      taskRef:
+        name: golang-test
+      workspaces:
+        - name: source
+          workspace: shared-workspace
+      params:
+        - name: package
+          value: "github.com/tektoncd/pipeline"
+        - name: packages
+          value: "./pkg/..."
+      matrix:
+        params:
+          - name: GOARCH
+            value:
+              - "linux/amd64"
+              - "linux/ppc64le"
+          - name: flags
+            value:
+              - "-race -cover -v"
+        include:
+          - name: s390x-no-race
+            value:
+              GOARCH: "linux/s390x"
+              flags: "-cover -v" 
+    - ...
+```
+
+Instead of the [the solution](#define-explicit-combinations-in-the-matrix) for including a specific combination in the
+`Matrix`, we could use `Parameter` of type `Object` as such:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: pipeline-to-build-images-from-a-single-repo
+spec:
+  workspaces:
+    - name: shared-workspace
+  tasks:
+    - ...
+    - name: kaniko-build
+      taskRef:
+        name: kaniko
+      workspaces:
+        - name: source
+          workspace: shared-workspace
+      matrix:
+        include:
```

Comment:
I was actually picturing this as an array of object params; not necessarily suggesting we go down that route but I think it's an option worth adding here

Reply:
what we have here is an array of object params 🙃

Reply:
or could you show me an example of what you mean?

Reply:
oh sorry-- I was thinking of something like this:

```yaml
apiVersion: tekton.dev/v1beta1
kind: Pipeline
metadata:
  name: pipeline-to-build-images-from-a-single-repo
spec:
  workspaces:
    - name: shared-workspace
  params:
  - name: build-config
     type: array
     default:
     - type: object
        properties:
          IMAGE: "image-1"
          DOCKERFILE: "path/to/Dockerfile1"
     - type: object
        properties:
              IMAGE: "image-2"
              DOCKERFILE: "path/to/Dockerfile2"
     - type: object
        properties:
              IMAGE: "image-2"
              DOCKERFILE: "path/to/Dockerfile2"
  tasks:
    - name: kaniko-build
      taskRef:
        name: kaniko
      workspaces:
        - name: source
          workspace: shared-workspace
      matrix:
      - name: build-config
         value: $(params.build-config)    
    - ...
```
and then the task could reference $(params.build-config.IMAGE) and $(params.build-config.DOCKERFILE)

Reply:
we don't support array of objects in the form you've described there, and would rather not go there 😅

Reply:
Yeah having written it out it's a bit awkward 😅